src/dune_lang/package_variable_name.ml

```diff
@@ -49,13 +42,65 @@ let version = of_string "version"
 let name = of_string "name"
 let build = of_string "build"
 let post = of_string "post"
-let one_of t xs = List.mem xs ~equal t
 let dev = of_string "dev"
+let one_of t xs = List.mem xs ~equal t
 
 let platform_specific =
   Set.of_list [ arch; os; os_version; os_distribution; os_family; sys_ocaml_version ]
 ;;
 
+(* Users are free to refer to arbitrarily-named variables, but these are the
+   common variables that are widely in circulation. *)
+let all_known =
+  [ arch
+  ; os
+  ; os_version
+  ; os_distribution
+  ; os_family
+  ; opam_version
+  ; sys_ocaml_version
+  ; with_test
+  ; with_doc
+  ; with_dev_setup
+  ; version
+  ; name
+  ; build
+  ; post
+  ; dev
+  ]
+;;
+
+let encode t = Encoder.string (to_string t)
+
+let check_typo_underscore_instead_of_dash loc t =
+  let string = to_string t in
+  if String.contains string '_'
+  then
+    List.iter all_known ~f:(fun v ->
+      (* Many of dune's pform variables use underscores to separate words, but
+         package variables use underscores for compatibility with opam. Thus we
+         expect users to often mistakenly use underscores for package variables,
+         so warn in this case. *)
+      let v_string = to_string v in
+      let possible_typo = String.replace_char v_string ~from:'-' ~to_:'_' in
```

Comment:
That's a constant value for each `v` that is declared in `all_known`, so you could just store it there once and avoid recomputing it.